### ISSUE TITLE
Investigate NTLM auth for WinRM connections 

### ISSUE THREAD

**Issue by koikonom** [hashicorp/terraform-provider-ad#34]
This should be a small one, but needs to be tested before releasing it:

### Potential Terraform Configuration
```
endpoint := winrm.NewEndpoint("localhost", 5985, false, false, nil, nil, nil, 0)

params := DefaultParameters
params.TransportDecorator = func() Transporter { return &ClientNTLM{} }

client, err := NewClientWithParameters(endpoint, "test", "test", params)
if err != nil {
	panic(err)
}
```
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jpatigny** [comment]
I would also be interested to test the kerberos authentication.

Some developed something for it : 

https://github.com/dpotapov/winrm-auth-krb5

**Comment by jpatigny** [comment]
Regarding the kerberos part, I actually came up with a different scenario.
I'm running terraform insider a docker container using a gMSA to authenticate to the AD.
In that case, I only need to pass cmdlet locally and I don't have to connect to any remote host using winrm.

Would it be considerable to implement the possibility to execute commands locally ?

**Comment by koikonom** [comment]
Hi @jpatigny, can you share a bit more information about your setup ?

I assume that by gMSA you refere to https://docs.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/group-managed-service-accounts-overview ?

**Comment by jpatigny** [comment]
@koikonom indeed.

First you need to setup all AD part (create gMSA , grant permissions etc) : 

https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts

Once this is done, you can run a docker container based on a windows image to interact with the active directory : 

https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/gmsa-run-container

Next, in order to make terraform interact with the AD, you need to build a docker terraform image based on windows OS.

Example of manual execution (on the docker host) :

docker run --security-opt "credentialspec=file://contoso_webapp01.json" --rm -it terraform-windows:latest get-aduser -identity myuser

Finally, I integrated the all thing with a CICD (gitlab runner using docker.
I registered a runner on the docker host server that will use the docker image  

https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runnersdocker-section

![image](https://user-images.githubusercontent.com/40454756/95510222-e1fb3780-09b5-11eb-8b35-c2d7e07208c6.png)
![image](https://user-images.githubusercontent.com/40454756/95510296-f93a2500-09b5-11eb-8ea7-23f6dca0f729.png)

I'll make an article some day that explain the all thing (if i'm motivated).

**Comment by koikonom** [comment]
Thanks for the detailed response. I'll give this a look and see what I can do :) !

**Comment by koikonom** [comment]
Hi @jpatigny, I gave this some more thought and I think it is doable but the changes I have in mind will require some extensive testing. The approach I have in mind is to alter the logic so if the provider runs on windows and there are no credentials configured then we should try running the command locally instead of going via WinRM.

I will probably give this a go in the release after the next one.

**Comment by jpatigny** [comment]
@koikonom ,

Looks to be a good approach to me.
It would really help because I'm currently using a custom provider that is clearly less evolved than this one and I would prefer to participate to this one ;-)

I'm looking forward to test it with new features ;-)

**Comment by jpatigny** [comment]
@koikonom ,

Is it still in your plan to implement the local execution of commands even if kerberos authentication has been implemented ?
I still need it because it's too complex to use winrm within the container (I would need to use a service account, then generate a kerberos ticket) 
I would loose the feature of using a gMSA to manage the AD.

I'll have a look anything to implement it but it's a lot of work.

**Comment by koikonom** [comment]
I still think it could be useful to do and it would solve a few problems, but I am not sure where it stands in terms of prioritization. 

I'm going to file it as a new issue and discuss with the team.

**Comment by koikonom** [comment]
Closing this issue since @czeumer's PR addressed the original issue. We can continue discussing the rest in #70 :)
